pyOCD/rtos/rtx.py

```diff
@@ -0,0 +1,338 @@
+        return sp
+
+    def get_stack_frame(self):
+        return self._target_context.read8(self._base + STACKFRAME_OFFSET)
+
+
+CURRENT_OFFSET = 20
+THREADLIST_OFFSET = 36
+DELAYLIST_OFFSET = 44
+WAITLIST_OFFSET = 48
+THREADNEXT_OFFSET = 8
+DELAYNEXT_OFFSET = 16
+
+## @brief Base class for RTOS support plugins.
+class RTXThreadProvider(ThreadProvider):
+    def __init__(self, target):
+        super(RTXThreadProvider, self).__init__(target)
+
+    def init(self, symbolProvider):
+        # Lookup required symbols.
+        self._osRtxInfo =symbolProvider.get_symbol_value('osRtxInfo')
+        if self._osRtxInfo is None:
+            return False
+        log.debug('init(), found osRtxInfo')
+        self._threads = {}
+        return True
+
+    def get_threads(self):
+        if not self.is_enabled:
+            return []
+        return list(self._threads.values())
+
+    def invalidate(self):
+        self._threads = {}
+
+    def event_handler(self, notification):
+        # Invalidate threads list if flash is reprogrammed.
+        self.invalidate();
+
+    def _build_thread_list(self):
+        log.debug('_build_thread_list()')
+        self._readylist = self._osRtxInfo + THREADLIST_OFFSET
+        self._delaylist = self._osRtxInfo + DELAYLIST_OFFSET
+        self._waitlist = self._osRtxInfo + WAITLIST_OFFSET
+
+        # Currently running Thread
+        thread = self._target_context.read32(self._osRtxInfo + CURRENT_OFFSET)
+        if thread:
+            log.debug('found RUNNING thread %x' % thread)
+            if not thread in self._threads:
+                self._threads[thread] = RTXTargetThread(self._target_context, self, thread)
+            self._current_id = thread
+            self._current = self._threads[thread]
+
+        # Scan Ready list
+        thread = self._target_context.read32(self._readylist)
+        while thread:
+            log.debug('found READY thread %x' % thread)
+            # targetContext, provider, base
+            if not thread in self._threads:
+                self._threads[thread] = RTXTargetThread(self._target_context, self, thread)
+            thread = self._target_context.read32(thread+THREADNEXT_OFFSET)
+
+        # Delay list
+        thread = self._target_context.read32(self._delaylist)
+        while thread:
+            log.debug('found DELAY thread %x' % thread)
+            if not thread in self._threads:
+                self._threads[thread] = RTXTargetThread(self._target_context, self, thread)
+            thread = self._target_context.read32(thread+DELAYNEXT_OFFSET)
+
+        # Wait list
+        thread = self._target_context.read32(self._waitlist)
+        while thread:
+            log.debug('found WAIT thread %x' % thread)
+            if not thread in self._threads:
+                self._threads[thread] = RTXTargetThread(self._target_context, self, thread)
+            thread = self._target_context.read32(thread+DELAYNEXT_OFFSET)
+
```

Comment:
You should probably add the code to create the Handler mode pseudo-thread here. This is important so that if you stop in an interrupt handler, it appears as a separate "thread".

Example code copied from Argon (same for FreeRTOS and Zephyr):
```py
        # Create fake handler mode thread.
        if self.get_ipsr() > 0:
            log.debug("creating handler mode thread")
            t = HandlerModeThread(self._target_context, self)
            newThreads[t.unique_id] = t
```